gnd/src/main/java/com/google/android/gnd/repository/DataRepository.java

```diff
@@ -112,23 +156,87 @@ private void onProjectLoaded(Project project) {
     return getFeature(projectId, featureId)
         .switchIfEmpty(Single.error(new DocumentNotFoundException()))
-        .flatMap(feature -> localDataStore.getRecords(feature))
-        .map(summaries -> filterSummariesByFormId(summaries, formId));
+        .flatMapPublisher(
+            feature -> {
+              Flowable<RemoteDataEvent<Record>> remoteChanges =
+                  remoteDataStore
+                      .loadRecordSummariesOnceAndStreamChanges(feature)
+                      .subscribeOn(Schedulers.io());
+              Flowable<ImmutableList<Record>> localChanges =
+                  localDataStore
+                      .getRecordsOnceAndStream(feature, formId)
+                      .subscribeOn(Schedulers.io());
+              return maybeSyncFirst(remoteChanges)
+                  .toFlowable()
+                  .map(o -> (ImmutableList<Record>) o)
+                  .concatWith(
+                      // This will rewrite the first update in remoteChanges to the local db
+                      // even if already successful. We allow this in case the first update
+                      // timed out or failed or the network wasn't available.
+                      localChanges.mergeWith(
+                          remoteChanges.flatMapCompletable(this::mergeRemoteRecordChange)));
+            });
   }
 
-  private List<Record> filterSummariesByFormId(List<Record> summaries, String formId) {
-    return stream(summaries)
-        .filter(record -> record.getForm().getId().equals(formId))
-        .collect(Collectors.toList());
+  /**
+   * If network if available, wait for first remote emission and update of local store then
+   * complete, otherwise complete immediately. Also completes with success if record sync fails or
+   * times out.
+   */
+  private Completable maybeSyncFirst(Flowable<RemoteDataEvent<Record>> remoteChanges) {
+    if (networkManager.isNetworkAvailable()) {
+      Log.d(TAG, "Has network; syncing records before showing");
+      return remoteChanges
+          .firstElement()
+          .flatMapCompletable(this::mergeRemoteRecordChange)
+          .timeout(GET_REMOTE_RECORDS_TIMEOUT_SECS, TimeUnit.SECONDS)
+          .doOnError(t -> Log.d(TAG, "Record sync timed out"))
+          .onErrorComplete();
+    } else {
+      Log.d(TAG, "No network; skipping remote sync");
+      return Completable.complete();
+    }
   }
 
+  private Completable mergeRemoteRecordChange(RemoteDataEvent<Record> event) {
+    switch (event.getEventType()) {
+      case ENTITY_LOADED:
+      case ENTITY_MODIFIED:
+        return event
+            .value()
+            .map(
+                record ->
+                    localDataStore
+                        .mergeRecord(record)
+                        .subscribeOn(Schedulers.io())
+                        .doOnError(e -> Log.e(TAG, "ERROR: ", e)))
+            .orElse(Completable.never());
```

Comment:
What are the consequences of reaching this? 

Will subscriptions to the completable in question simply not fire?

Reply:
Looks like I forgot this in while debugging; `doOnError` doesn't modify the stream, so it wouldn't have had any effect. When an error does happen, it get propagated in the stream up to the `RecordListViewModel`, which will eventually show an error.